src/components/Pagination/Pagination.tsx

```diff
@@ -1,45 +1,199 @@
 
-export const Pagination = () => {
-  const { page, setPage, totalPages } = usePagination()
+export type PaginationProps = ButtonGroupProps & {
+  maxButtons?: number | false
+  buttonProps?: ButtonProps
+}
+
+const createButton = (page: number, currentPage: number, props: ButtonProps) => (
+  <Button key={page} isActive={currentPage === page} {...props}>
+    {page + 1}
+  </Button>
+)
+
+const createEllipsisButton = (
+  key: string,
+  onClick: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void,
+  showInput: boolean,
+  onKeyDown: (e: React.KeyboardEvent<HTMLInputElement>) => void,
+  onBlur: () => void
+) => {
+  if (showInput) {
+    return <Input key={key} placeholder='Page #' width='50px' onKeyDown={onKeyDown} onBlur={onBlur} autoFocus />
+  } else {
+    return (
+      <Button key={key} as='a' href='#goto-page' onClick={onClick}>
+        ...
+      </Button>
+    )
+  }
+}
+
+const usePaginationPages = (
+  currentPage: number,
+  totalPages: number | undefined,
+  maxButtons: number | undefined | false,
+  gotoPage: (page: number) => void,
+  createPageButton: (i: number) => ReactElement
+) => {
+  const [startEllipsisInput, setStartEllipsisInput] = useState(false)
+  const [endEllipsisInput, setEndEllipsisInput] = useState(false)
 
-  const pages: ReactElement[] = useMemo(() => {
-    const pages: ReactElement[] = []
+  return useMemo(() => {
+    if (totalPages === undefined) return []
+
+    let pages: ReactElement[] = []
+
+    // Create an array of all page buttons
     for (let i = 0; i < totalPages; i++) {
-      pages.push(
-        <Button key={i} onClick={() => setPage(i)} isActive={page === i}>
-          {i + 1}
-        </Button>
-      )
+      pages.push(createPageButton(i))
     }
-    return pages
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [page, totalPages])
 
-  return <ButtonGroup isAttached>{pages.map((page) => page)}</ButtonGroup>
+    if (!maxButtons || totalPages <= maxButtons) {
+      return pages
+    }
+
+    const startEllipsis = createEllipsisButton(
+      'start-ellipsis',
+      (e) => {
+        e.preventDefault()
+        setStartEllipsisInput(true)
+      },
+      startEllipsisInput,
+      (e) => {
+        if (e.key === 'Enter') {
+          const pageNumber = Number((e.target as HTMLInputElement).value)
+          gotoPage(pageNumber - 1)
+          setStartEllipsisInput(false)
+        }
+      },
+      () => setStartEllipsisInput(false)
+    )
+
+    const endEllipsis = createEllipsisButton(
+      'end-ellipsis',
+      (e) => {
+        e.preventDefault()
+        setEndEllipsisInput(true)
+      },
+      endEllipsisInput,
+      (e) => {
+        if (e.key === 'Enter') {
+          const pageNumber = Number((e.target as HTMLInputElement).value)
+          gotoPage(pageNumber - 1)
+          setEndEllipsisInput(false)
+        }
+      },
+      () => setEndEllipsisInput(false)
+    )
```

Comment:
This code is duplicated. Why just don't store the ellipsis input state inside the component and pass the `gotoPage` as parameter of the component?

```ts
const EllipsisButton = ({ key, gotoPage }: { key: string; gotoPage: (page: number) => void }) => {
  const [ellipsisInput, setEllipsisInput] = useState(false)

  if (ellipsisInput) {
    return (
      <Input
        key={key}
        placeholder='Page #'
        width='50px'
        onKeyDown={(e) => {
          if (e.key === 'Enter') {
            const pageNumber = Number((e.target as HTMLInputElement).value)
            gotoPage(pageNumber - 1)
            setEllipsisInput(false)
          }
        }}
        onBlur={() => setEllipsisInput(false)}
        autoFocus
      />
    )
  } else {
    return (
      <Button
        key={key}
        as='a'
        href='#goto-page'
        onClick={(e) => {
          e.preventDefault()
          setEllipsisInput(true)
        }}
      >
        ...
      </Button>
    )
  }
}
```

And then just:

```ts
    const startEllipsis = <EllipsisButton key={'start-ellipsis'} gotoPage={gotoPage} />
    const endEllipsis = <EllipsisButton key={'end-ellipsis'} gotoPage={gotoPage} />
```

Or similar. Provably you could use just use one `EllipsisButton` component

Reply:
True, it ended up this way after lots of iterations of the code. But I'll change it.